### ISSUE TITLE
[rl] qrdqn/quantile_loss.rs: shape/panic guards, Kappa/taus API tightening, test gaps

### ISSUE THREAD

**Issue by anthonytorlucci** [anthonytorlucci/rlevo#440]
- **Files:** `algorithms/qrdqn/quantile_loss.rs`
- **Body:** Re-review of `quantile_huber_loss_per_sample` after the ADR-0050-§14 refactor (unreduced per-sample return, `&Tensor` reference args). The κ=0→inf defect itself is tracked separately (#345); these are the remaining hardening/ergonomics/coverage gaps against the current file:
  1. **Loop-invariant hoist** (`:104`): `target_quantiles.clone().unsqueeze_dim::<3>(1)` runs on every chunk iteration inside the `while` loop even though `target_quantiles` never changes across chunks — for `N=200, CHUNK=32` that's ~7 redundant unsqueeze+clone passes per `learn_step`. Hoist the `unsqueeze_dim` outside the loop (the per-iteration `.clone()` is still required since Burn's `-` consumes operands, but the metadata computation itself is one-time).
  2. **Silent panic on shape mismatch** (`:96-119`): if `target_quantiles`'s or `taus`'s dims don't match `pred_quantiles`, the broadcast/subtract inside the loop panics via Burn's own shape assertions with no attribution to this function. The `# Panics` doc section (`:75-78`) only documents the `n_pred == 0` case, not shape mismatches. At minimum, document the mismatch-panics-not-errors contract; ideally validate `target_quantiles.dims()` and `taus.dims()` up front with a named panic message.
  3. **`.expect()` on empty accumulator** (`:130`): `per_sample_acc.expect(...)` panics on `num_quantiles == 0`. The agent always configures `num_quantiles >= 1`, so this is defense-in-depth rather than an observed bug, but a `pub` numerics fn callable from benchmarks/experiments panicking mid-training-loop on bad input is worth a named `assert!` guard at the top (or a `Result` return, which is a breaking change — scope separately).
  4. **`taus` bounds unchecked**: nothing enforces `taus ∈ [0,1]`, so a caller passing midpoints for the wrong `N`, or non-midpoint values, gets a numerically valid but theoretically meaningless loss with no signal. A `debug_assert!` over the host-read values would catch this in tests/benches at zero release cost.
  5. **API surface wider than used**: `taus: Tensor<B, 1>` duplicates information derivable from `pred_quantiles.dims()[1]` (the midpoint convention `τ_i = (i+0.5)/N`) — same root cause as #186's `quantile_taus`-rebuild issue, but at the loss-function signature itself rather than the agent's call site. Consider deriving `taus` internally (breaking change) or at minimum validating `taus.dims()[0] == n_pred`.
  6. **`kappa: f32` admits illegal states**: complementary to #345's config-layer fix — a `Kappa` newtype validated at construction (rejecting `<= 0.0`/non-finite) would let the loss function itself drop its runtime check. Breaking change; sequence after #345's minimal fix lands.
  7. **Argument order** diverges from `c51::loss` (`pred, target, ...]` here vs. `target, pred` there) — cosmetic but visible on two sibling `pub` loss functions in the same crate.
  8. **`huber` is `pub` but has no cross-crate consumer**: grepped `rlevo-benchmarks` for `quantile_loss::huber`/`quantile_huber_loss` — no hits. `huber` is only used internally (by this file) and in its own tests; narrow to `pub(crate)`.
  9. **Test coverage gaps**: no test for `kappa <= 0` (the exact defect #345 tracks — a regression test here would pin the guard once added), no test for `n_pred == 0` (the `:130` panic), no property-based test for non-negativity/`1/κ`-scaling (only single hand-picked examples exist today), no pinned aggregation-order reference test (`mean_j → sum_i → mean_batch` is easy to silently invert to `sum`/`mean`).
- **Not included / already handled elsewhere:**
  - `QUANTILE_CHUNK_SIZE` (`:44`) is **not** `pub` — it has no visibility modifier at all, which in Rust means module-private regardless of the enclosing module's own `pub`-ness. A prior review draft's claim that it "follows the module visibility" is incorrect for the current file; no action needed.
  - `pred_quantiles.clone()` before `.slice()` (`:96-98`) is **not removable**: `Tensor::slice(self, ...)` in burn-tensor 0.21.0 (`burn-tensor-0.21.0/src/tensor/api/base.rs:1352`) consumes `self` by value, and this function now takes `pred_quantiles: &Tensor<B, 2>` by reference, so the clone is required to obtain an owned tensor to slice. No change needed.
- **Severity:** No High — the one High-severity item (κ=0 → inf) is tracked by #345. These are hardening/ergonomics/coverage gaps on an otherwise numerically-correct, already-well-tested loss function.
- **Source:** Code review 2026-06-30, §1.2, §2.1, §3.1, §3.2, §3.3, §5.1, §5.2, §7.1, §7.2, §7.3, §7.4, §8.1

**Comment by anthonytorlucci** [comment]
Re-checked against #345, which landed today. #345 validates κ at the **config boundary only** — `config::positive` + an `is_finite` guard in `QrDqnTrainingConfig::validate` (`qrdqn_config.rs:214-228`). `quantile_loss.rs` is byte-for-byte unchanged, so items 1, 5 and 7 are unaffected by construction.

**Item 6 (`Kappa` newtype) — the rationale changes, and not in the direction I first expected.**

The *config-path* half is now dead. Every `QrDqnTrainingConfig` construction site reaches `validate()`: the only struct-literal site is `train.rs:204` (`..QrDqnTrainingConfig::default()`), which is passed to `QrDqnAgent::new`, and `new` calls `config.validate()?` before anything else (`qrdqn_agent.rs:178`); every builder site goes through `build()`, which also validates (`qrdqn_config.rs:409`). There is no bypass, so "a newtype is needed to stop an unvalidated κ reaching the loss *via a config*" is closed by #345.

**But the loss function is public API, and #345 does not protect it.**

`quantile_huber_loss_per_sample` is `pub fn` (`quantile_loss.rs:79`) inside `pub mod quantile_loss` → `pub mod qrdqn` → `pub mod algorithms` (`lib.rs:71-78`). So this compiles for any downstream user of the crate:

```rust
use rlevo_reinforcement_learning::algorithms::qrdqn::quantile_loss::quantile_huber_loss_per_sample;
let loss = quantile_huber_loss_per_sample(&pred, &target, &taus, 0.0); // -> NaN
```

The parameter has no `assert`, no `debug_assert`, and no documented precondition — the doc comment has a `# Panics` section for `num_quantiles == 0` but says nothing about κ. That is a real public entry point with a live NaN path, not a hypothetical one. The only in-repo direct caller is `benches/qrdqn_bench.rs:62`, which hardcodes a safe `kappa = 1.0`, and there are zero `rlevo-benchmarks*` consumers — so nothing in this repo is *currently* broken by it. The exposure is to external callers.

Two corrections to how this item is worded:

1. It says a newtype would "let the loss function itself drop its runtime check." There is no runtime check to drop — the loss validates κ not at all. The newtype would be *adding* the first guarantee, not relocating an existing one.
2. The cheapest fix is not necessarily the newtype. A `debug_assert!(kappa.is_finite() && kappa > 0.0)` plus a `# Panics`/preconditions doc line closes the documentation gap immediately and costs nothing in release builds. The `Kappa` newtype is the stronger version and would make it a compile-time property, at the cost of a breaking signature change.

Severity unchanged (this issue already says "No High") — nothing in the repo is miscomputing today. But the justification is now "an unguarded public numerics API", which is more concrete than the defense-in-depth framing it was filed under.

Executed evidence for the κ values themselves is on #345: κ = 0 yields **NaN**, not `inf` (`huber()` returns exactly `0.0` at κ = 0, so the division is `0/0`), and κ = `+∞` yields NaN in every element via `0 · (−∞)` in the eagerly-evaluated masked-out linear branch.
